### PR TITLE
Multilevel mailbox structure

### LINKED ISSUE #794: Sub-sub-folders should be expandable and hyperlinks from subfolders which are collapsed are clickable
### Steps to reproduce
1. Open Mail app

### Expected behaviour
All subfodlers and subsubfolders should be expandable.

### Actual behaviour
There are hyperlinks from subfolders which are collapsed that are clickable. Furthermore sub and subsub folders are not expandable they are shown as foo.bar or foo2.bar2.foo3 etc.

### Mail app

**Mail app version:** (see apps admin page)
0.7.10

**Mailserver or service:**
Courier

**Number of accounts:**
1

### Server configuration
**Operating system**:
Debain
**Web server:**
Apache 
**Database:**
Mysql
**PHP version:**
5
**Version:** (see admin page)

**Updated from an older version or fresh install:**
Updated
**List of activated apps:**
Mail, Contacts, Calendar

#### Client configuration
**Browser:**
Safari, Chome
**Operating system:**
Mac, Windows

##### Browser log
<details>
```
JQMIGRATE: Migrate is installed, version 1.4.0
build.js?v=bab260c7-4:91 Starting Mail …
DevTools failed to parse SourceMap: https://XXX/core/vendor/purify.min.js.map
build.js?v=bab260c7-4:24 GET https://XXX/index.php/apps/mail/api/avatars/url/SupportCenter%40ieee.org 404 (Not Found)
send @ build.js?v=bab260c7-4:24
ajax @ build.js?v=bab260c7-4:24
e @ build.js?v=bab260c7-4:162
n @ build.js?v=bab260c7-4:24
l._callHandler @ build.js?v=bab260c7-4:57
request @ build.js?v=bab260c7-4:57
_fetchAvatar @ build.js?v=bab260c7-4:138
onRender @ build.js?v=bab260c7-4:138
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
addChildView @ build.js?v=bab260c7-4:35
_addChild @ build.js?v=bab260c7-4:35
x.each.x.forEach @ build.js?v=bab260c7-4:24
_showCollection @ build.js?v=bab260c7-4:35
_renderChildren @ build.js?v=bab260c7-4:35
renderChildren @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
onRender @ build.js?v=bab260c7-4:101
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
showFolderContent @ build.js?v=bab260c7-4:98
y @ build.js?v=bab260c7-4:35
v @ build.js?v=bab260c7-4:35
d @ build.js?v=bab260c7-4:35
c.trigger @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
Promise resolved (async)
i @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:24
setTimeout (async)
x.delay @ build.js?v=bab260c7-4:24
O @ build.js?v=bab260c7-4:24
n @ build.js?v=bab260c7-4:24
(anonymous) @ build.js?v=bab260c7-4:57
s @ build.js?v=bab260c7-4:57
showFolder @ build.js?v=bab260c7-4:162
default @ build.js?v=bab260c7-4:162
execute @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
x.some.x.any @ build.js?v=bab260c7-4:24
loadUrl @ build.js?v=bab260c7-4:35
start @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
build.js?v=bab260c7-4:24 GET https://XXX/index.php/apps/mail/api/avatars/url/pfarramt.waldmohr%40evkirchepfalz.de 404 (Not Found)
send @ build.js?v=bab260c7-4:24
ajax @ build.js?v=bab260c7-4:24
e @ build.js?v=bab260c7-4:162
n @ build.js?v=bab260c7-4:24
l._callHandler @ build.js?v=bab260c7-4:57
request @ build.js?v=bab260c7-4:57
_fetchAvatar @ build.js?v=bab260c7-4:138
onRender @ build.js?v=bab260c7-4:138
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
addChildView @ build.js?v=bab260c7-4:35
_addChild @ build.js?v=bab260c7-4:35
x.each.x.forEach @ build.js?v=bab260c7-4:24
_showCollection @ build.js?v=bab260c7-4:35
_renderChildren @ build.js?v=bab260c7-4:35
renderChildren @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
onRender @ build.js?v=bab260c7-4:101
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
showFolderContent @ build.js?v=bab260c7-4:98
y @ build.js?v=bab260c7-4:35
v @ build.js?v=bab260c7-4:35
d @ build.js?v=bab260c7-4:35
c.trigger @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
Promise resolved (async)
i @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:24
setTimeout (async)
x.delay @ build.js?v=bab260c7-4:24
O @ build.js?v=bab260c7-4:24
n @ build.js?v=bab260c7-4:24
(anonymous) @ build.js?v=bab260c7-4:57
s @ build.js?v=bab260c7-4:57
showFolder @ build.js?v=bab260c7-4:162
default @ build.js?v=bab260c7-4:162
execute @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
x.some.x.any @ build.js?v=bab260c7-4:24
loadUrl @ build.js?v=bab260c7-4:35
start @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
build.js?v=bab260c7-4:24 GET https://XXX/index.php/apps/mail/api/avatars/url/mario.barbareschi%40unina.it 404 (Not Found)
send @ build.js?v=bab260c7-4:24
ajax @ build.js?v=bab260c7-4:24
e @ build.js?v=bab260c7-4:162
n @ build.js?v=bab260c7-4:24
l._callHandler @ build.js?v=bab260c7-4:57
request @ build.js?v=bab260c7-4:57
_fetchAvatar @ build.js?v=bab260c7-4:138
onRender @ build.js?v=bab260c7-4:138
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
addChildView @ build.js?v=bab260c7-4:35
_addChild @ build.js?v=bab260c7-4:35
x.each.x.forEach @ build.js?v=bab260c7-4:24
_showCollection @ build.js?v=bab260c7-4:35
_renderChildren @ build.js?v=bab260c7-4:35
renderChildren @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
onRender @ build.js?v=bab260c7-4:101
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
showFolderContent @ build.js?v=bab260c7-4:98
y @ build.js?v=bab260c7-4:35
v @ build.js?v=bab260c7-4:35
d @ build.js?v=bab260c7-4:35
c.trigger @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
Promise resolved (async)
i @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:24
setTimeout (async)
x.delay @ build.js?v=bab260c7-4:24
O @ build.js?v=bab260c7-4:24
n @ build.js?v=bab260c7-4:24
(anonymous) @ build.js?v=bab260c7-4:57
s @ build.js?v=bab260c7-4:57
showFolder @ build.js?v=bab260c7-4:162
default @ build.js?v=bab260c7-4:162
execute @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
x.some.x.any @ build.js?v=bab260c7-4:24
loadUrl @ build.js?v=bab260c7-4:35
start @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
build.js?v=bab260c7-4:24 GET https://XXX/index.php/apps/mail/api/avatars/url/iscas%40epapers.org 404 (Not Found)
send @ build.js?v=bab260c7-4:24
ajax @ build.js?v=bab260c7-4:24
e @ build.js?v=bab260c7-4:162
n @ build.js?v=bab260c7-4:24
l._callHandler @ build.js?v=bab260c7-4:57
request @ build.js?v=bab260c7-4:57
_fetchAvatar @ build.js?v=bab260c7-4:138
onRender @ build.js?v=bab260c7-4:138
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
addChildView @ build.js?v=bab260c7-4:35
_addChild @ build.js?v=bab260c7-4:35
x.each.x.forEach @ build.js?v=bab260c7-4:24
_showCollection @ build.js?v=bab260c7-4:35
_renderChildren @ build.js?v=bab260c7-4:35
renderChildren @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
onRender @ build.js?v=bab260c7-4:101
i @ build.js?v=bab260c7-4:35
triggerMethod @ build.js?v=bab260c7-4:35
render @ build.js?v=bab260c7-4:35
N @ build.js?v=bab260c7-4:35
show @ build.js?v=bab260c7-4:35
showChildView @ build.js?v=bab260c7-4:35
showFolderContent @ build.js?v=bab260c7-4:98
y @ build.js?v=bab260c7-4:35
v @ build.js?v=bab260c7-4:35
d @ build.js?v=bab260c7-4:35
c.trigger @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
Promise resolved (async)
i @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:57
(anonymous) @ build.js?v=bab260c7-4:24
setTimeout (async)
x.delay @ build.js?v=bab260c7-4:24
O @ build.js?v=bab260c7-4:24
n @ build.js?v=bab260c7-4:24
(anonymous) @ build.js?v=bab260c7-4:57
s @ build.js?v=bab260c7-4:57
showFolder @ build.js?v=bab260c7-4:162
default @ build.js?v=bab260c7-4:162
execute @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:35
x.some.x.any @ build.js?v=bab260c7-4:24
loadUrl @ build.js?v=bab260c7-4:35
start @ build.js?v=bab260c7-4:35
(anonymous) @ build.js?v=bab260c7-4:57
```
</details>

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/55307908-sub-sub-folders-should-be-expandable-and-hyperlinks-from-subfolders-which-are-collapsed-are-clickable?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by dehnhardt**
- [ ] (Soft) requires #3466
- [x] Logic for `addAccount`
- [x] Logic for `addMailbox`
- [x] ~~Edge case: removal of a mailbox that has children -> children should be promoted by one level (if IMAP even allow this operation)~~ -> problematic on IMAP level as well, so we're hiding the *delete* action for mailboxes with children.
- [x] Detect and persist personal namespace (related to https://github.com/nextcloud/mail/issues/3272, but fully automated)

Fixes #794 

Signed-off-by: Holger Dehnhardt <holger@dehnhardt.org>
With this pr it is possible to display a folder structure of any depth.
The structure is only created using the getter in the frontend. 
Creating new folders within the structure works - but is not tested carefully.
There are some problems with the display. 
* Replacing the folder icons with the arrows does not work properly 
image

* the navigation area quickly becomes too narrow.
See comments inside the changes...
image

_To make the goal of this PR clearer, this picture shows how the structure is presented so far:_
image

### CODE REVIEW

File: lib/Folder.php [hunk 1]
```diff
@@ -62,6 +68,9 @@ public function __construct(int $accountId, Horde_Imap_Client_Mailbox $mailbox,
 		$this->folders = [];
 		$this->status = [];
 		$this->specialUse = [];
+		$lastSeparator = strrpos($mailbox->utf8, $delimiter) > 0 ? strrpos($mailbox->utf8, $delimiter)+1: 0;
+		$this->displayName = substr($mailbox->utf8, $lastSeparator);
+		$this->path = substr($mailbox->utf8, 0, $lastSeparator);
 	}
```
Comment (dehnhardt) [inline lib/Folder.php:74]: Seperate display name and path. The path is used to select the subfolders.

File: src/store/getters.js [hunk 2]
```diff
@@ -38,12 +38,16 @@ export const getters = {
...
+		return state.accounts[accountId].folders
+			.map((folderId) => state.folders[folderId])
+			.filter((folder) => folder.path === '')
 	},
-	getSubfolders: (state, getters) => (accountId, folderId) => {
-		const folder = getters.getFolder(accountId, folderId)
-
-		return folder.folders.map((id) => state.folders[id])
+	getSubfolders: (state) => (accountId, folderId) => {
+		const folder = state.folders[normalizedFolderId(accountId, folderId)]
```
Comment (dehnhardt) [inline src/store/getters.js:46]: getters.getFolder(accountId, folderId) does not work here.

File: src/store/mutations.js [hunk 3]
```diff
@@ -56,7 +56,7 @@ export default {
 		)
 
 		// Save the folders to the store, but only keep IDs in the account's folder list
-		const folders = buildMailboxHierarchy(sortMailboxes(account.folders || []), havePrefix(account.folders))
+		const folders = sortMailboxes(account.folders || [])
```
Comment (dehnhardt) [inline src/store/mutations.js:59]: We don't need building a hierarchy here anymore.

**Comment by jancborchardt** [comment]
Looks quite nice! To prevent the issue of the navigation getting too narrow, we need to do a less extreme indent for the subentries. That is, the left side of the icon should left-align with the text of the parent folder.

**Comment by dehnhardt** [comment]
@jancborchardt I had the same idea. But I wasn't sure if there are design guidelines from Nextcloud that prohibit this kind of changes.
Would the change you proposed have to be made by someone specific to the Nextcloud core, as it could affect many apps?

**Comment by dehnhardt** [comment]
I opened an issue in https://github.com/nextcloud/nextcloud-vue for the display problems.
https://github.com/nextcloud/nextcloud-vue/issues/1140

**Comment by dehnhardt** [comment]
@ChristophWurst, @GretaD: May I draw your attention again to this ticket for three reasons. 
- the bug #3289 would have been fixed already
- the bug #794 should be fixed - as well as the requirement to display the tree structure correctly
- To implement the Sieve filter interface I need a mailbox picker, which I would like to implement with the correct tree structure.
However, I would like to make sure that the structure as I implemented it is ok.
I know that the last weeks were quite busy because of the release, but maybe you can find time now for a short evaluation.

File: lib/Folder.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/Folder.php:74]: we only need it on the front-end. can we just generate this there?

File: src/components/NavigationFolder.vue [hunk 4]
```diff
@@ -33,7 +33,7 @@
 	>
 		<!-- actions -->
 		<template slot="actions">
-			<template v-if="top">
+			<template>
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply src/components/NavigationFolder.vue:36]: fixed by #3290

File: src/components/NavigationFolder.vue [hunk 5]
```diff
@@ -175,6 +175,7 @@ export default {
 		},
 		subFolders() {
 			return this.$store.getters.getSubfolders(this.account.id, this.folder.id)
+			//return this.folder.folders
```
Comment (ChristophWurst) [inline src/components/NavigationFolder.vue:178]: remove?

File: lib/Folder.php (hunk 1 above)
Comment (dehnhardt) [reply lib/Folder.php:74]: Yes, we can;-), separating path and file name and creating the hierarchy on the fly is ok so far?

File: src/components/NavigationFolder.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (dehnhardt) [reply src/components/NavigationFolder.vue:36]: I know. But I was first;-)

File: lib/Folder.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/Folder.php:74]: I'm not entirely sure. But I did not fully grasp the intended change so far

File: src/components/NavigationFolder.vue (hunk 5 above)
Comment (dehnhardt) [reply src/components/NavigationFolder.vue]: Both done...

**Comment by ChristophWurst** [comment]
I'm currently focusing on #2125 but I'll look into this soon.

**Comment by dehnhardt** [comment]
> I'm currently focusing on #2125 but I'll look into this soon.

I've seen that. Take your time ;-)

**Comment by dehnhardt** [comment]
@ChristophWurst: Now that https://github.com/nextcloud/nextcloud-vue/issues/1140 has come to the mails masterbranch, wouldn't it be wonderful to have this in master as well ;-)
Rebased and squashed.

**Comment by ChristophWurst** [comment]
Sure, let's do this :) This requires a few adaptations for #3439 but I can help with that if you like.

File: lib/Folder.php [hunk 6]
```diff
@@ -45,6 +45,13 @@ class Folder {
 	/** @var string[] */
 	private $specialUse;
 
+	/** @var string */
+	private $path;
```
Comment (ChristophWurst) [inline lib/Folder.php:49]: there is no code that write to this, so it will always remain `null`

**Comment by ChristophWurst** [comment]
I took the liberty to just rebase the branch, resolve the conflicts and fix some unused code (leftover from previous rebase possibly).

The patch looks clean, but I'm unsure if your attended changes are still here. Could you please check and give me a quick update? We can always reset to an older state if I accidentally dropped some important changes :pray:

**Comment by ChristophWurst** [comment]
Bildschirmfoto von 2020-08-27 14-25-04
7d5a79a7afb55fe9f7b593a9465cc3f5


woooooooorks again :)

**Comment by dehnhardt** [comment]
To be honest, I can no longer tell which changes were originally mine and which are yours.
But I checked out and tested the branch on my test system and it seems to work how it should.
I just wonder why the indentation is still so big. Is the current Nextcloud vue version not merged yet?

**Comment by ChristophWurst** [comment]
> To be honest, I can no longer tell which changes were originally mine and which are yours.
> But I checked out and tested the branch on my test system and it seems to work how it should.

No worries. I spent some time on this yesterday and also added some unit tests.

> I just wonder why the indentation is still so big. Is the current Nextcloud vue version not merged yet?

It's not, hence the TODO in the PR description.

**Review by GretaD — APPROVED** [review]
tested and works

**Pull request merged**